rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+there's only on such slot pointing to one given value at all time, even if this
+slot is shared by many closures.
+
+Compared to the traditional closure conversion of Haskell, it seems that the
+advantage is to avoid copying as much as possible (given Lua's "large" value
+representation). First, many closures can share the same slot, instead of each
+having their own copy of the captured value. Second, I suppose that if the
+enclosing function outlives the closure, no value is ever moved out from the
+slot, and thus no copying happens.
+
+On the other hand, this techniques looks useless if the values are represented
+on one word, as a pointer to the slot and the slot itself still need to be
+allocated, plus the machinery around that (list of open closures during
+compilation and at runtime to decide what should be moved when de-allocating a
+frame).
+
+#### Virtual machine
+
+The Lua virtual machine is register-based since Lua 5.0 (note that there's still
+a stack for activation records) while most bytecode VMs out there are usually
+stack-based, even if they might have a few registers. Lua developers argue that
+a register-based VM saves many `push` and `pop` operations, which both avoids a
+non-trivial amount of copying (especially with the "large" tagged union
+representation of values) and reduces the number of instructions (albeit the
+instructions are larger, because they often need additional operands to indicate
+which registers they operate on). Registers also reduce the number of
+instructions needed: the Lua VM only has 35. The Lua VM has 256 registers
+represented as a packed array at runtime, guaranteeing fast access.
+
+#### References
+
+- [The Implementation of LUA 5](https://www.lua.org/doc/jucs05.pdf)
+
+### OCaml
+
+#### Memory representation
+
+The following notes on the memory representation applies to the native code
+backend's representation. I'm not sure how closures are represented in the Zinc
+Abstract Machine.
+
+OCaml uses a uniform memory representation where any value is represented as a
+single machine word. Unboxed values (integers, etc.) are distinguished from
+pointers by their least significant bit (and are thus encoded on `n-1` bits
+compared to their, say, C equivalent). This is needed for garbage collection
+only.
+
+Boxed values are represented as a pointer to a block, which is a contiguous area
+of the memory with a one-word header:
+
+The header holds:
+
+- the size
+- the color (for garbage-collection)
+- a multi-purpose tag byte
+
+The header is followed by arbitrary content whose shape depends on the type of
+the value.
+
+```text
+| header (1w)        | content   |
+----------------------------------
+| size | color | tag | data ...  |
+```
+
+ADTs are represented within the block's data as an integer for constructors with
+no argument (the tag byte then doesn't store the actual contructor's tag but has
+the same value than for a boxed `int`). For a variant with parameters, the tag
```

Comment:
This is not quite it. A constructor without argument is represented as an _unboxed_ integer value. It doesn't point to a block, so there is no tag involved. Constructors with arguments are pointers, and point to a structure as above.

Reply:
You're right, I don't know where I got this idea